src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs

```diff
@@ -795,118 +791,85 @@ public string EncryptToken(string innerJwt, EncryptingCredentials encryptingCred
-                    if (!string.IsNullOrEmpty(compressionAlgorithm))
-                    {
-                        try
-                        {
-                            plainText = CompressToken(innerJwt, compressionAlgorithm);
-                        }
-                        catch (Exception ex)
-                        {
-                            throw LogHelper.LogExceptionMessage(new SecurityTokenCompressionFailedException(LogHelper.FormatInvariant(TokenLogMessages.IDX10680, compressionAlgorithm), ex));
-                        }
-                    }
-                    else
-                    {
-                        plainText = Encoding.UTF8.GetBytes(innerJwt);
-                    }
+                var header = CreateDefaultJWEHeader(encryptingCredentials, compressionAlgorithm, tokenType);
+                if (additionalHeaderClaims != null)
+                    header.Merge(JObject.FromObject(additionalHeaderClaims));
 
+                byte[] plainText;
+                if (!string.IsNullOrEmpty(compressionAlgorithm))
+                {
                     try
                     {
-                        var rawHeader = Base64UrlEncoder.Encode(Encoding.UTF8.GetBytes(header.ToString(Formatting.None)));
-                        var encryptionResult = encryptionProvider.Encrypt(plainText, Encoding.ASCII.GetBytes(rawHeader));
-                        return string.Join(".", rawHeader, Base64UrlEncoder.Encode(wrappedKey), Base64UrlEncoder.Encode(encryptionResult.IV), Base64UrlEncoder.Encode(encryptionResult.Ciphertext), Base64UrlEncoder.Encode(encryptionResult.AuthenticationTag));
+                        plainText = CompressToken(innerJwt, compressionAlgorithm);
                     }
                     catch (Exception ex)
                     {
-                        throw LogHelper.LogExceptionMessage(new SecurityTokenEncryptionFailedException(LogHelper.FormatInvariant(TokenLogMessages.IDX10616, encryptingCredentials.Enc, encryptingCredentials.Key), ex));
+                        throw LogHelper.LogExceptionMessage(new SecurityTokenCompressionFailedException(LogHelper.FormatInvariant(TokenLogMessages.IDX10680, compressionAlgorithm), ex));
                     }
                 }
+                else
+                {
+                    plainText = Encoding.UTF8.GetBytes(innerJwt);
+                }
+
+                try
+                {
+                    var rawHeader = Base64UrlEncoder.Encode(Encoding.UTF8.GetBytes(header.ToString(Formatting.None)));
+                    var encryptionResult = encryptionProvider.Encrypt(plainText, Encoding.ASCII.GetBytes(rawHeader));
+                    return JwtConstants.DirectKeyUseAlg.Equals(encryptingCredentials.Alg, StringComparison.Ordinal)? string.Join(".", rawHeader, string.Empty, Base64UrlEncoder.Encode(encryptionResult.IV), Base64UrlEncoder.Encode(encryptionResult.Ciphertext), Base64UrlEncoder.Encode(encryptionResult.AuthenticationTag)) :
```

Comment:
nit: consider spliting this line to increase readability.

Reply:
Done.